homeassistant/components/knx/__init__.py

```diff
@@ -171,35 +175,80 @@
 
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up the KNX integration."""
+    """Start the KNX integration."""
+    conf: ConfigType | None = config.get(DOMAIN)
+
+    if conf is None:
+        # If we have a config entry, setup is done by that config entry.
+        # If there is no config entry, this should fail.
+        return bool(hass.config_entries.async_entries(DOMAIN))
+
+    conf = dict(conf)
+
+    hass.data[DATA_KNX_CONFIG] = conf
+
+    # Only import if we haven't before.
+    if not hass.config_entries.async_entries(DOMAIN):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": config_entries.SOURCE_IMPORT}, data=conf
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Load a config entry."""
+    conf = hass.data.get(DATA_KNX_CONFIG)
+
+    #  When reloading
+    if conf is None:
+        conf = await async_integration_yaml_config(hass, DOMAIN)
+        if not conf or DOMAIN not in conf:
+            return False
+
+        conf = conf[DOMAIN]
+        hass.data[DATA_KNX_CONFIG] = conf
+
+    # If user didn't have configuration.yaml config, generate defaults
+    if conf is None:
+        conf = CONFIG_SCHEMA({DOMAIN: dict(entry.data)})[DOMAIN]
+
+    config = {**entry.data, **conf}
+
     try:
-        knx_module = KNXModule(hass, config)
-        hass.data[DOMAIN] = knx_module
+        knx_module = KNXModule(hass, config, entry)
         await knx_module.start()
     except XKNXException as ex:
-        _LOGGER.warning("Could not connect to KNX interface: %s", ex)
-        hass.components.persistent_notification.async_create(
-            f"Could not connect to KNX interface: <br><b>{ex}</b>", title="KNX"
-        )
+        raise ConfigEntryNotReady from ex
+
+    hass.data[DOMAIN] = knx_module
 
-    if CONF_KNX_EXPOSE in config[DOMAIN]:
-        for expose_config in config[DOMAIN][CONF_KNX_EXPOSE]:
+    if CONF_KNX_EXPOSE in config:
+        for expose_config in config[CONF_KNX_EXPOSE]:
             knx_module.exposures.append(
                 create_knx_exposure(hass, knx_module.xknx, expose_config)
             )
 
-    for platform in SupportedPlatforms:
-        if platform.value not in config[DOMAIN]:
-            continue
+    async def setup_platforms() -> None:
+        """Set up platforms."""
+        await asyncio.gather(
+            *(
+                hass.config_entries.async_forward_entry_setup(entry, platform.value)
+                for platform in SupportedPlatforms
+                if platform.value in config
+            )
+        )
+
+    hass.async_create_task(setup_platforms())
+
+    # set up notify platform, no entry support for notify component yet,
```

Comment:
TODO: Use config entry once https://github.com/home-assistant/core/commits/add-service-integration-foundation is merged.